README.md
# rgeoSrv
[![](https://img.shields.io/github/workflow/status/sams96/rgeoSrv/continuous-integration?style=for-the-badge)](https://github.com/sams96/rgeoSrv/actions?query=workflow%3Acontinuous-integration)
[![](https://goreportcard.com/badge/github.com/sams96/rgeoSrv?style=for-the-badge)](https://goreportcard.com/report/github.com/sams96/rgeoSrv)
[![Release](https://img.shields.io/github/tag/sams96/rgeoSrv.svg?label=release&color=24B898&logo=github&style=for-the-badge)](https://github.com/sams96/rgeoSrv/releases/latest)
[![go.dev reference](https://img.shields.io/badge/go.dev-reference-007d9c?logo=go&logoColor=white&style=for-the-badge)](https://pkg.go.dev/github.com/sams96/rgeoSrv)

rgeoSrv wraps the package rgeo into a reverse geocoding microservice.

See [github.com/sams96/rgeo](https://github.com/sams96/rgeo) for more
information on rgeo.

### Installation

    go get github.com/sams96/rgeoSrv/..

or,

    docker pull docker.pkg.github.com/sams96/rgeosrv/rgeosrv

### Usage

    rgeoSrv -addr localhost:8080

or,

	docker run -p 8080:8080 docker.pkg.github.com/sams96/rgeosrv/rgeosrv and

then:

	curl "localhost:8080/query?0&51.5045"

will return:

	{"country":"United Kingdom","country_long":"United Kingdom of Great Britain and Northern Ireland","country_code_2":"GB","country_code_3":"GBR","continent":"Europe","region":"Europe","subregion":"Northern Europe","province":"Tower Hamlets","province_code":"GB-TWH","city":"London"}


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.0.1] - 2020-02-24

### Initial release
 - Wraps the library [rgeo](https://github.com/sams96/rgeo) in a http server
   with JSON responses
 - Docker build published with releases from now on


main_test.go
package main

import (
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/go-test/deep"
)

func TestRootHandler(t *testing.T) {
	tests := []struct {
		name           string
		in             *http.Request
		out            *httptest.ResponseRecorder
		expectedStatus int
		expectedBody   string
	}{
		{
			name:           "Algeria",
			in:             httptest.NewRequest("GET", "/?1.880273&31.787305", nil),
			out:            httptest.NewRecorder(),
			expectedStatus: http.StatusOK,
			expectedBody:   `{"Country":"Algeria","CountryLong":"People's Democratic Republic of Algeria","CountryCode2":"DZ","CountryCode3":"DZA","Continent":"Africa","Region":"Africa","SubRegion":"Northern Africa"}`,
		},
		{
			name:           "Madagascar",
			in:             httptest.NewRequest("GET", "/?47.478275&-17.530126", nil),
			out:            httptest.NewRecorder(),
			expectedStatus: http.StatusOK,
			expectedBody:   `{"Country":"Madagascar","CountryLong":"Republic of Madagascar","CountryCode2":"MG","CountryCode3":"MDG","Continent":"Africa","Region":"Africa","SubRegion":"Eastern Africa"}`,
		},
		{
			name:           "Zimbabwe",
			in:             httptest.NewRequest("GET", "/?29.832875&-19.948725", nil),
			out:            httptest.NewRecorder(),
			expectedStatus: http.StatusOK,
			expectedBody:   `{"Country":"Zimbabwe","CountryLong":"Republic of Zimbabwe","CountryCode2":"ZW","CountryCode3":"ZWE","Continent":"Africa","Region":"Africa","SubRegion":"Eastern Africa"}`,
		},
		{
			name:           "Ocean",
			in:             httptest.NewRequest("GET", "/?0&0", nil),
			out:            httptest.NewRecorder(),
			expectedStatus: http.StatusInternalServerError,
			expectedBody:   "country not found\n",
		},
		{
			name:           "North Pole",
			in:             httptest.NewRequest("GET", "/?-135&90", nil),
			out:            httptest.NewRecorder(),
			expectedStatus: http.StatusInternalServerError,
			expectedBody:   "country not found\n",
		},
		{
			name:           "South Pole",
			in:             httptest.NewRequest("GET", "/?44.99&-89.99", nil),
			out:            httptest.NewRecorder(),
			expectedStatus: http.StatusOK,
			expectedBody:   `{"Country":"Antarctica","CountryLong":"","CountryCode2":"AQ","CountryCode3":"ATA","Continent":"Antarctica","Region":"Antarctica","SubRegion":"Antarctica"}`,
		},
		{
			name:           "Alaska",
			in:             httptest.NewRequest("GET", "/?-150.542&66.3", nil),
			out:            httptest.NewRecorder(),
			expectedStatus: http.StatusOK,
			expectedBody:   `{"Country":"United States of America","CountryLong":"United States of America","CountryCode2":"US","CountryCode3":"USA","Continent":"North America","Region":"Americas","SubRegion":"Northern America"}`,
		},
		{
			name:           "UK",
			in:             httptest.NewRequest("GET", "/?0&52", nil),
			out:            httptest.NewRecorder(),
			expectedStatus: http.StatusOK,
			expectedBody:   `{"Country":"United Kingdom","CountryLong":"United Kingdom of Great Britain and Northern Ireland","CountryCode2":"GB","CountryCode3":"GBR","Continent":"Europe","Region":"Europe","SubRegion":"Northern Europe"}`,
		},
		{
			name:           "Libya",
			in:             httptest.NewRequest("GET", "/?24.994611&25.860750", nil),
			out:            httptest.NewRecorder(),
			expectedStatus: http.StatusOK,
			expectedBody:   `{"Country":"Libya","CountryLong":"Libya","CountryCode2":"LY","CountryCode3":"LBY","Continent":"Africa","Region":"Africa","SubRegion":"Northern Africa"}`,
		},
		{
			name:           "Egypt",
			in:             httptest.NewRequest("GET", "/?25.005187&25.855963", nil),
			out:            httptest.NewRecorder(),
			expectedStatus: http.StatusOK,
			expectedBody:   `{"Country":"Egypt","CountryLong":"Arab Republic of Egypt","CountryCode2":"EG","CountryCode3":"EGY","Continent":"Africa","Region":"Africa","SubRegion":"Northern Africa"}`,
		},
		{
			name:           "US Border",
			in:             httptest.NewRequest("GET", "/?-102.560616&48.998074", nil),
			out:            httptest.NewRecorder(),
			expectedStatus: http.StatusOK,
			expectedBody:   `{"Country":"United States of America","CountryLong":"United States of America","CountryCode2":"US","CountryCode3":"USA","Continent":"North America","Region":"Americas","SubRegion":"Northern America"}`,
		},
		{
			name:           "Canada Border",
			in:             httptest.NewRequest("GET", "/?-102.560616&49.0", nil),
			out:            httptest.NewRecorder(),
			expectedStatus: http.StatusOK,
			expectedBody:   `{"Country":"Canada","CountryLong":"Canada","CountryCode2":"CA","CountryCode3":"CAN","Continent":"North America","Region":"Americas","SubRegion":"Northern America"}`,
		},
	}

	for _, test := range tests {
		test := test
		t.Run(test.name, func(t *testing.T) {
			rootHandler(test.out, test.in)
			if test.out.Code != test.expectedStatus {
				t.Errorf("expected: %d\ngot: %d\n",
					test.expectedStatus, test.out.Code)
				t.Fail()
			}

			body := test.out.Body.String()
			if diff := deep.Equal(test.expectedBody, body); diff != nil {
				t.Error(diff)
			}
		})
	}
}


query/query_test.go
/*
Copyright 2020 Sam Smith

Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License.  You may obtain a copy of the
License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed
under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
CONDITIONS OF ANY KIND, either express or implied.  See the License for the
specific language governing permissions and limitations under the License.
*/

package query

import (
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/sergi/go-diff/diffmatchpatch"
)

func TestQuery(t *testing.T) {
	tests := []struct {
		name           string
		in             *http.Request
		out            *httptest.ResponseRecorder
		expectedStatus int
		expectedBody   string
	}{
		{
			name:           "Algeria",
			in:             httptest.NewRequest("GET", "/query?1.880273&31.787305", nil),
			out:            httptest.NewRecorder(),
			expectedStatus: http.StatusOK,
			expectedBody:   `{"country":"Algeria","country_long":"People's Democratic Republic of Algeria","country_code_2":"DZ","country_code_3":"DZA","continent":"Africa","region":"Africa","subregion":"Northern Africa","province":"El Bayadh","province_code":"DZ-32"}`,
		},
		{
			name:           "Madagascar",
			in:             httptest.NewRequest("GET", "/query?47.523836&-18.905691", nil),
			out:            httptest.NewRecorder(),
			expectedStatus: http.StatusOK,
			expectedBody:   `{"country":"Madagascar","country_long":"Republic of Madagascar","country_code_2":"MG","country_code_3":"MDG","continent":"Africa","region":"Africa","subregion":"Eastern Africa","province":"Analamanga","province_code":"MG-T","city":"Antananarivo"}`,
		},
		{
			name:           "Zimbabwe",
			in:             httptest.NewRequest("GET", "/query?29.832875&-19.948725", nil),
			out:            httptest.NewRecorder(),
			expectedStatus: http.StatusOK,
			expectedBody:   `{"country":"Zimbabwe","country_long":"Republic of Zimbabwe","country_code_2":"ZW","country_code_3":"ZWE","continent":"Africa","region":"Africa","subregion":"Eastern Africa","province":"Midlands","province_code":"ZW-MI"}`,
		},
		{
			name:           "Ocean",
			in:             httptest.NewRequest("GET", "/query?0&0", nil),
			out:            httptest.NewRecorder(),
			expectedStatus: http.StatusInternalServerError,
			expectedBody:   "country not found\n",
		},
		{
			name:           "North Pole",
			in:             httptest.NewRequest("GET", "/query?-135&90", nil),
			out:            httptest.NewRecorder(),
			expectedStatus: http.StatusInternalServerError,
			expectedBody:   "country not found\n",
		},
		{
			name:           "South Pole",
			in:             httptest.NewRequest("GET", "/query?44.99&-89.99", nil),
			out:            httptest.NewRecorder(),
			expectedStatus: http.StatusOK,
			expectedBody:   `{"country":"Antarctica","country_code_2":"AQ","country_code_3":"ATA","continent":"Antarctica","region":"Antarctica","subregion":"Antarctica","province":"Antarctica","province_code":"AQ-X01~"}`,
		},
		{
			name:           "Alaska",
			in:             httptest.NewRequest("GET", "/query?-149.901785&61.199134", nil),
			out:            httptest.NewRecorder(),
			expectedStatus: http.StatusOK,
			expectedBody:   `{"country":"United States of America","country_long":"United States of America","country_code_2":"US","country_code_3":"USA","continent":"North America","region":"Americas","subregion":"Northern America","province":"Alaska","province_code":"US-AK","city":"Anchorage"}`,
		},
		{
			name:           "UK",
			in:             httptest.NewRequest("GET", "/query?0&51.5045", nil),
			out:            httptest.NewRecorder(),
			expectedStatus: http.StatusOK,
			expectedBody:   `{"country":"United Kingdom","country_long":"United Kingdom of Great Britain and Northern Ireland","country_code_2":"GB","country_code_3":"GBR","continent":"Europe","region":"Europe","subregion":"Northern Europe","province":"Tower Hamlets","province_code":"GB-TWH","city":"London"}`,
		},
		{
			name:           "Libya",
			in:             httptest.NewRequest("GET", "/query?24.98&25.86", nil),
			out:            httptest.NewRecorder(),
			expectedStatus: http.StatusOK,
			expectedBody:   `{"country":"Libya","country_long":"Libya","country_code_2":"LY","country_code_3":"LBY","continent":"Africa","region":"Africa","subregion":"Northern Africa","province":"Al Kufrah","province_code":"LY-KF"}`,
		},
		{
			name:           "Egypt",
			in:             httptest.NewRequest("GET", "/query?25.005187&25.855963", nil),
			out:            httptest.NewRecorder(),
			expectedStatus: http.StatusOK,
			expectedBody:   `{"country":"Egypt","country_long":"Arab Republic of Egypt","country_code_2":"EG","country_code_3":"EGY","continent":"Africa","region":"Africa","subregion":"Northern Africa","province":"Al Wadi at Jadid","province_code":"EG-WAD"}`,
		},
		{
			name:           "US Border",
			in:             httptest.NewRequest("GET", "/query?-102.560616&48.992073", nil),
			out:            httptest.NewRecorder(),
			expectedStatus: http.StatusOK,
			expectedBody:   `{"country":"United States of America","country_long":"United States of America","country_code_2":"US","country_code_3":"USA","continent":"North America","region":"Americas","subregion":"Northern America","province":"North Dakota","province_code":"US-ND"}`,
		},
		{
			name:           "Canada Border",
			in:             httptest.NewRequest("GET", "/query?-102.560616&49.0", nil),
			out:            httptest.NewRecorder(),
			expectedStatus: http.StatusOK,
			expectedBody:   `{"country":"Canada","country_long":"Canada","country_code_2":"CA","country_code_3":"CAN","continent":"North America","region":"Americas","subregion":"Northern America","province":"Saskatchewan","province_code":"CA-SK"}`,
		},
	}

	q, err := NewHandlers(nil)
	if err != nil {
		t.Error(err)
	}

	dmp := diffmatchpatch.New()

	for _, test := range tests {
		test := test
		t.Run(test.name, func(t *testing.T) {
			q.query(test.out, test.in)
			if test.out.Code != test.expectedStatus {
				t.Errorf("expected: %d\ngot: %d\n",
					test.expectedStatus, test.out.Code)
				t.Fail()
			}

			body := test.out.Body.String()
			if test.expectedBody != body {
				diff := dmp.DiffMain(test.expectedBody, body, false)
				t.Error(dmp.DiffPrettyText(diff))
			}
		})
	}
}


query/query.go
/*
Copyright 2020 Sam Smith

Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License.  You may obtain a copy of the
License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed
under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
CONDITIONS OF ANY KIND, either express or implied.  See the License for the
specific language governing permissions and limitations under the License.
*/

/*
Package query implements the query handler for rgeoSrv
*/
package query

import (
	"encoding/json"
	"errors"
	"log"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"

	"github.com/sams96/rgeo"
)

// Handlers supplies dependencies for query handlers
type Handlers struct {
	l *log.Logger
	r *rgeo.Rgeo
}

// NewHandlers creates new Handlers type for query
func NewHandlers(l *log.Logger) (*Handlers, error) {
	r, err := rgeo.New(rgeo.Provinces10, rgeo.Cities10)
	if err != nil {
		return nil, err
	}

	return &Handlers{
		l: l,
		r: r,
	}, nil
}

// query handles request to "/query"
func (h *Handlers) query(w http.ResponseWriter, r *http.Request) {
	coords, err := parseURL(r.URL)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	loc, err := h.r.ReverseGeocode(coords)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	resp, err := json.Marshal(loc)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/json; charset=utf-8")
	w.WriteHeader(http.StatusOK)

	_, err = w.Write(resp)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
}

// parseURL extracts the coordinates from the request URL
func parseURL(u *url.URL) ([]float64, error) {
	if u.RawQuery == "" || !strings.ContainsRune(u.RawQuery, '&') {
		return []float64{},
			errors.New("rgeoSrv requires a request in the form /query?lon&lat")
	}

	coord := strings.Split(u.RawQuery, "&")

	lon, err := strconv.ParseFloat(coord[0], 64)
	if err != nil {
		return []float64{}, err
	}

	lat, err := strconv.ParseFloat(coord[1], 64)
	if err != nil {
		return []float64{}, err
	}

	return []float64{lon, lat}, nil
}

// logger is http handler middleware which adds logging
func (h *Handlers) logger(next http.HandlerFunc) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		startTime := time.Now()
		next(w, r)
		h.l.Printf("Request for %s from %s processed in %s\n",
			r.URL.Path, r.RemoteAddr, time.Since(startTime))
	}
}

// SetupRoutes sets up the query routes on the given mux
func (h *Handlers) SetupRoutes(mux *http.ServeMux) {
	mux.HandleFunc("/query", h.logger(h.query))
}


main.go
/*
Copyright 2020 Sam Smith

Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License.  You may obtain a copy of the
License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed
under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
CONDITIONS OF ANY KIND, either express or implied.  See the License for the
specific language governing permissions and limitations under the License.
*/

/*
Command rgeoSrv wraps the package rgeo into a reverse geocoding microservice.

See https://github.com/sams96/rgeo for more information on rgeo.

Installation

	go get github.com/sams96/rgeoSrv/..
or,
	docker pull docker.pkg.github.com/sams96/rgeosrv/rgeosrv

Usage

	rgeoSrv -addr localhost:8080
or,
	docker run -p 8080:8080 docker.pkg.github.com/sams96/rgeosrv/rgeosrv
and then:
	curl "localhost:8080/query?0&51.5045"
will return:
	{"country":"United Kingdom","country_long":"United Kingdom of Great Britain and Northern Ireland","country_code_2":"GB","country_code_3":"GBR","continent":"Europe","region":"Europe","subregion":"Northern Europe","province":"Tower Hamlets","province_code":"GB-TWH","city":"London"}
*/
package main

import (
	"flag"
	"log"
	"net/http"
	"os"
	"time"

	"github.com/sams96/rgeoSrv/query"
)

var (
	defaultSrvAddr = ":8080"
)

func main() {
	// Bind flags
	srvAddr := flag.String("addr", defaultSrvAddr, "Address to open the server on")
	flag.Parse()

	// Create logger
	l := log.New(os.Stderr, "rgeoSrv ", log.LstdFlags)

	// Create mux and add handlers
	mux := http.NewServeMux()
	q, err := query.NewHandlers(l)
	if err != nil {
		l.Fatalf("server failed to start: %v", err)
	}

	q.SetupRoutes(mux)

	// Create and start server
	srv := newServer(mux, *srvAddr)

	l.Println("server starting at", *srvAddr)
	err = srv.ListenAndServe()
	if err != nil {
		l.Fatalf("server failed to start: %v", err)
	}
}

// newServer creates a new http server
func newServer(mux *http.ServeMux, addr string) *http.Server {
	srv := &http.Server{
		Addr:         addr,
		ReadTimeout:  5 * time.Second,
		WriteTimeout: 10 * time.Second,
		IdleTimeout:  120 * time.Second,
		Handler:      mux,
	}

	return srv
}
